flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -1181,137 +1184,284 @@ struct EmboxOpConversion : public FIROpConversion<fir::EmboxOp> {
     auto *thisBlock = rewriter.getInsertionBlock();
     auto func = mlir::cast<mlir::LLVM::LLVMFuncOp>(thisBlock->getParentOp());
     rewriter.setInsertionPointToStart(&func.front());
-    auto sz = genConstantOffset(loc, rewriter, 1);
+    auto sz = this->genConstantOffset(loc, rewriter, 1);
     auto al = rewriter.create<mlir::LLVM::AllocaOp>(loc, toTy, sz, alignment);
     rewriter.restoreInsertionPoint(thisPt);
     return al;
   }
 
-  mlir::LLVM::BitcastOp genGEPToField(mlir::Location loc,
-                                      mlir::LLVM::LLVMType ty,
-                                      mlir::ConversionPatternRewriter &rewriter,
-                                      mlir::Value base, mlir::Value zero,
-                                      int field) const {
-    auto fld = genConstantOffset(loc, rewriter, field);
-    auto gep = genGEP(loc, ty, rewriter, base, zero, fld);
-    return rewriter.create<mlir::LLVM::BitcastOp>(loc, ty, gep);
+  template <typename... FLDS>
+  mlir::LLVM::GEPOp genGEPToField(mlir::Location loc, mlir::LLVM::LLVMType ty,
+                                  mlir::ConversionPatternRewriter &rewriter,
+                                  mlir::Value base, mlir::Value zero,
+                                  FLDS... fields) const {
+    return this->genGEP(loc, ty.getPointerTo(), rewriter, base, zero,
+                        this->genConstantOffset(loc, rewriter, fields)...);
+  }
+
+  static mlir::LLVM::LLVMType getBoxEleTy(mlir::LLVM::LLVMType boxPtrTy,
+                                          unsigned i) {
+    return boxPtrTy.getPointerElementTy().getStructElementType(i);
+  }
+
+  // Perform an extension or truncation as needed on an integer value
+  mlir::Value integerCast(mlir::Location loc,
+                          mlir::ConversionPatternRewriter &rewriter,
+                          mlir::LLVM::LLVMType ty, mlir::Value val) const {
+    auto toSize = ty.getUnderlyingType()->getPrimitiveSizeInBits();
+    auto fromSize = val.getType()
+                        .cast<mlir::LLVM::LLVMType>()
+                        .getUnderlyingType()
+                        ->getPrimitiveSizeInBits();
+    if (toSize < fromSize)
+      return rewriter.create<mlir::LLVM::TruncOp>(loc, ty, val);
+    if (toSize > fromSize)
+      return rewriter.create<mlir::LLVM::SExtOp>(loc, ty, val);
+    return val;
+  }
+
+  // Get the element size and CFI type code of the boxed value.
+  std::tuple<mlir::Value, mlir::Value>
+  getSizeAndTypeCode(mlir::Location loc,
+                     mlir::ConversionPatternRewriter &rewriter,
+                     mlir::Type boxEleTy) const {
```

Comment:
Have you consider using `Fortran::runtime::TypeCode(category, kind)` here ? https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/runtime/type-code.h
I understand it may not ideal because it uses category/kind which I do not how how hard it is to retrieve here. But that would allow having a single point of truth for this mapping.

Reply:
I did. Unfortunately, it's interface is wrong for the reason stated.